Fix: Adding collection form model validation errors to parent form

Dhamsoft opened the pull request:
- Extract validation methods to own module
    - Add #run_backing_model_validation method to validate backing model allowing with collection backing model validation
    - Add flag variable and logic to run the validation with or without backing_model validation
    - Reorder the callback lifecycle
      The backing model before validation and validation moved above form before_save and around_save

Commit:
Add test coverage for validation fails on the child collection records

Commit:
Fix add collection forms validation to parent form

Commit:
Adding collection form errors logic moveto CollectionWrapper class

- Update OnForm::Form.expose_collection_of to validate collection forms
  parent form object passed to CollectionWrapper to add errors back to parent form

- Update CollectionWrapper#validate_forms to receive parent_form as argument
  and add logic to add errors to parent if any

- Add test to check form validation errors added back to parent form

Commit:
Adding collection form errors logic move to CollectionWrapper class

- Extract validation methods to own module
- Add #run_backing_model_validation method to validate backing model allowing with collection backing model validation
- Add flag variable and logic to run the validation with or without backing_model validation
- Reorder the callback lifecycle
  The backing model before validation and validation moved above form before_save and around_save

Dhamsoft commented:
@willbryant  Alternate approach for https://github.com/fluxfederation/on_form/pull/3

binary-koan commented on lib/on_form/saving.rb:
```diff
@@ -14,26 +14,21 @@ def transaction(&block)
+        self.backing_model_validations = true
+        run_backing_model_validation!
+        raise ActiveModel::ValidationError, self if has_form_errors
+        raise ActiveRecord::RecordInvalid, self unless errors.empty?
```
We got into an interesting discussion here about `ValidationError` vs `RecordInvalid`, and whether there's a meaningful distinction between custom validations failing and model validations failing.

We also talked about when ActiveRecord does particular callbacks in the saving process: it explicitly does validation (with hooks) THEN save (with hooks), so it will happily let you make your model invalid during a before_save callback. We wondered whether it would make sense to mirror this here, so run validations for the model + form, then run save callbacks and validate without saving.

willbryant replied:
IMHO yes there is a meaningful distinction, simply because you wouldn't expect a non-record class to raise a RecordInvalid (what would it set #record to?).

I think if we have to choose, we should usually be aiming to be compatible with ActiveRecord's callback order, unless there is something fatally wrong with it or it is impossible to implement, as it's what people will expect as they refactor fat models into forms.  But I don't think I fully implemented it that way because I think it was difficult to achieve without creating other problems.

BTW, did you notice that the callback order is actually specified in the README?  That would need to be changed if we proceed with this behavior change (and a major version bump, but I expect we'll be doing that anyway).  The readme currently says:

> Note that model save calls are nested inside the form save calls, which means that although form validation takes place before form save starts, model validation takes place after form saving begins.

You mention the ActiveRecord behavior in the case you have both validation callbacks and save callbacks.  What happens in the case where an associated model has both callbacks itself?  Do both the original and associated model before_validations run before the before_save on the original record?

willbryant commented on lib/on_form/saving.rb:
```diff
@@ -14,26 +14,21 @@ def transaction(&block)
-            collect_errors
-            raise
-          end
+          backing_model_instances.each { |backing_model| backing_model.save(validate: false) }
```
What happens now if a before_save callback on the backing model saves another model with a save! call?

binary-koan replied:
Interesting point. I guess my intuitive idea of how it would work is

* validations (+ callbacks) on the form object
  * validations (+ callbacks) on directly associated models
    * validations (+ callbacks) on nested models

then assuming these pass ...

* in a transaction
  * save callbacks on the form objects
    * save callbacks on directly associated models
      * save callbacks on nested models

So if a before_save callback on the backing model calls `save!` and it explodes, the transaction would roll back and nothing would be saved.

Not sure how well that maps to what ActiveRecord does.

willbryant commented on lib/on_form/saving.rb:
```diff
@@ -14,26 +14,21 @@ def transaction(&block)
+
+        self.backing_model_validations = false
+        has_form_errors = !run_validations!
+        self.backing_model_validations = true
```
Not thrilled about this at all.  At a minimum, it would need to be in a try/ensure block.

Could you explain a bit more about why it is needed?

Dhamsoft replied:
This is setting form object level. If someone updates attributes of form object and chack `valid?` it will exclude backing model validation because we set false while saving. To make sure we are resetting here.

Commit:
Remove duplicate validation module and update collection form validation and save logic

- Remove collection form saving via after_save callback
- Update save with saving collection form with validate false
- Update run_validations with validating collection forms
- Update test case to cover both form and backing model validation on collection

willbryant commented:
Beautiful!
